src/clr-core/internal/base/focus-trap.base.ts

```diff
@@ -0,0 +1,117 @@
+  private previousFocus: Element;
+  private topReboundEl: any;
+  private bottomReboundEl: any;
+  private onFocusInEvent: any;
+
+  @property({ attribute: false })
+  private hostElement = this;
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    if (document.activeElement) {
+      this.previousFocus = document.activeElement;
+    }
+
+    // add class to prevent scroll on body
+    document.body.classList.add('no-scrolling');
+
+    this.addReboundElements();
+
+    FocusTrapTracker.setCurrent(this.hostElement);
+    this.hostElement.setAttribute('tabindex', '0');
+    this.hostElement.focus();
+
+    this.onFocusInEvent = this.onFocusIn.bind(this);
+    document.addEventListener('focusin', this.onFocusInEvent);
+  }
+
+  disconnectedCallback() {
+    document.removeEventListener('focusin', this.onFocusInEvent);
+
+    // remove class to prevent scroll on body
+    document.body.classList.remove('no-scrolling');
+
+    this.removeReboundElements();
+    FocusTrapTracker.activatePreviousCurrent();
+
+    if (isHTMLElement(this.previousFocus)) {
+      (this.previousFocus as HTMLElement).focus();
+    }
+
+    super.disconnectedCallback();
+  }
+
+  protected onFocusIn($event: any) {
+    if (FocusTrapTracker.getCurrent() !== this.hostElement) {
+      return;
+    }
+
+    if ($event.target && this.hostElement) {
+      if (
+        !this.hostElement.contains($event.target) ||
+        $event.target === this.topReboundEl ||
+        $event.target === this.bottomReboundEl
+      ) {
+        this.hostElement.focus();
+      }
+    }
+  }
+
+  protected createFocusableOffScreenEl(): any {
+    if (isBrowser()) {
+      const offScreenSpan = document.createElement('span');
+      offScreenSpan.setAttribute('tabindex', '0');
+      offScreenSpan.classList.add('offscreen-focus-rebounder');
+      return offScreenSpan;
+    }
+  }
+
+  protected addReboundElements() {
+    if (isBrowser()) {
+      this.topReboundEl = this.createFocusableOffScreenEl();
+      this.bottomReboundEl = this.createFocusableOffScreenEl();
+      if (this.parentElement) {
+        this.parentElement.insertBefore(this.topReboundEl, this.hostElement);
+        if (this.nextSibling) {
+          this.parentElement.insertBefore(this.bottomReboundEl, this.nextSibling);
+        } else {
+          this.parentElement.appendChild(this.bottomReboundEl);
+        }
+      }
+    }
+  }
+
+  protected removeReboundElements() {
+    if (isBrowser()) {
+      if (this.topReboundEl && this.parentElement) {
+        this.parentElement.removeChild(this.topReboundEl);
+      }
+      if (this.bottomReboundEl && this.parentElement) {
+        this.parentElement.removeChild(this.bottomReboundEl);
+      }
+
+      // These are here to to make sure that
+      // we completely delete all traces of the removed DOM objects.
+      delete this.topReboundEl;
+      delete this.bottomReboundEl;
+    }
+  }
+
+  protected render() {
+    return html`<slot></slot>`;
+  }
+}
```

Comment:
Does the focus-trap have unit tests?

Reply:
Same logic as the button base, we test it out by testing the components that extend it.

Reply:
I'm not a huge fan of testing everything through integration, although that should maybe be a deeper discussion outside of this context.

I'd rather know that the focus-trap is breaking and be able to zero in quickly to fix it than see that a change I made broke the datepicker, wizard, modal, combobox, signposts, and dropdowns and have to do the hop-skip-jump to realize what I really broke was the focus-trap.

When I test the modal, I care about things that are specific to the modal.

We should discuss this with @coryrylan at some point because we're going to be frequently stomping on each other's toes if we have philosophically opposed views on unit testing.